### ISSUE TITLE
[BUG]:  cannot use generic type ristretto.Cache[K z.Key, V any] without instantiation

### ISSUE THREAD

**Issue by poornachandratejasvi** [dgraph-io/ristretto#394]
### What version of Ristretto are you using?

# github.com/ory/viper
/root/go/pkg/mod/github.com/ory/viper@v1.7.5/viper.go:224:16: cannot use generic type ristretto.Cache[K z.Key, V any] without instantiation
getting below error post ristretto go package is updated to **v1.0.0** from  github.com/dgraph-io/ristretto **v0.1.1** 

/root/go/pkg/mod/github.com/ory/viper@v1.7.5/viper.go:253:17: implicit function instantiation requires go1.18 or later (-lang was set to go1.12; check go.mod)
/root/go/pkg/mod/github.com/ory/viper@v1.7.5/viper.go:253:37: cannot use generic type ristretto.Config[K z.Key, V any] without instantiation
/root/go/pkg/mod/github.com/ory/viper@v1.7.5/viper.go:307:15: cannot use generic type ristretto.Cache[K z.Key, V any] without instantiation
/root/go/pkg/mod/github.com/ory/viper@v1.7.5/viper.go:307:36: cannot use generic type ristretto.Config[K z.Key, V any] without instantiation

### What version of Go are you using?

1.22.6

### Have you tried reproducing the issue with the latest release?

None

### What is the hardware spec (RAM, CPU, OS)?

docker arm64

### What steps will reproduce the bug?

any package who is using this package will get the issue 

github.com/ory/go-acc

**Comment by jney** [comment]
Same issue using badger and upgrading:

```diff
- github.com/dgraph-io/ristretto v0.1.1 // indirect
+ github.com/dgraph-io/ristretto v1.0.0 // indirect
```

**Comment by OK93-01-18** [comment]
i have same issue on go v1.23.1

**Comment by Hamper** [comment]
The repository at github.com/ory/viper was archived 4 years ago. The current version of github.com/spf13/viper does not use ristretto.

Ristretto v1 uses generics and may not be compatible with packages that require ristretto without generics. You can try using ristretto v0 (the latest version is v0.2.0).

**Comment by jney** [comment]
@Hamper I don't use viper, but I use badger

**Comment by Hamper** [comment]
Badger supports ristretto with generics starting from v4.3.0. Versions v4.2.0 and below do not support generics in the call to newCache and require ristretto v0

**Comment by jchappelow** [comment]
The related issue in the badger repo: https://github.com/dgraph-io/badger/issues/2005

This is tricky to resolve if you have an app with direct or transitive dependencies such that you include both badger/v3 and badger/v4.  It feels like the only way to resolve this is if ristretto were to have gone from v0.x to v2 so that the module import path for each version of ristretto would be different (two different `requires` for the incompatible ristrettos), allowing badger/v3 to keep using v0.1 or v0.2 while badger/v3 could use ristretto/v2.  Right now using badger/v4 hoist the ristretto requirement and breaks badger/v3.

I'm in a situation where our dependencies are forcing the ristretto version up, thus breaking our badger/v3 import.

**Comment by mangalaman93** [comment]
We just did a badger release https://github.com/dgraph-io/badger/releases/tag/v4.3.1 and ristretto has been out for a couple of weeks now https://github.com/dgraph-io/ristretto/releases/tag/v1.0.0. Please check if you are still facing issues and feel free to reopen this GitHub issue. Thanks

**Comment by mangalaman93** [comment]
Could you help me understand the dependency graph that leads to both badger v3 and v4 in the dependency tree?

**Comment by carsonip** [comment]
+1 on what @jchappelow raised. Since risteretto v0 and v1 share the same import path, it is impossible to have badgerv2 and badgerv4 in the same project as they use v0 and v1 respectively, but go mod will pick v1, causing badgerv2 code fail to compile. If risteretto releases v2 and badgerv4 uses risteretto v2, then badgerv2 can keep using risteretto v0 and badgerv4 use risteretto v2.

**Comment by carsonip** [comment]
> Is it possible that we can upgrade the badger version in the dependencies that still depend on older version of badger?

We have a setup that contains both v2 and v4 badger, but that does not compile as mentioned. Upgrading the v2 part to v4 is not straightforward as there are on-disk changes and require migration.

> I am also going to look at the possibility of releasing ristretto v2.

If there's a ristretto v2, and a new badger v4.3.2 that uses ristretto v2, that should fix it, as it would allow a v2 badger to use ristretto v0.1.1, and v4 badger to use ristretto v2 independently.

**Comment by carsonip** [comment]
Also, an affected project doesn't have to be using v2 and v4 badger directly. It could be a v2 badger user, but at the same time uses jaeger. Looking at [jaeger](https://github.com/jaegertracing/jaeger/blob/main/go.mod), it pulls in badger v4.3.1, which then pulls in ristretto v1.0.0, which causes badger v2 code to not compile.

Go versioning assumes v0 provides no compatibility guarantees, therefore it does not require an import path change between v0 and v1. But when a breaking change is really released in a library between v0 and v1, it means code that relies on v0 library will be broken if v1 library is pulled in. That's why it is better to release breaking changes only in v2.

**Comment by mangalaman93** [comment]
I understand, thanks for the explanation. Is it possible that we can upgrade the badger version in the dependencies that still depend on older version of badger? I am also going to look at the possibility of releasing ristretto v2. Will that help?

**Comment by carsonip** [comment]
> I understand, thanks for the explanation. Is it possible that we can upgrade the badger version in the dependencies that still depend on older version of badger? I am also going to look at the possibility of releasing ristretto v2. Will that help?

@mangalaman93 that seems like a duplicate comment. Did you mean to type something else?

**Comment by mangalaman93** [comment]
> > I understand, thanks for the explanation. Is it possible that we can upgrade the badger version in the dependencies that still depend on older version of badger? I am also going to look at the possibility of releasing ristretto v2. Will that help?
> 
> @mangalaman93 that seems like a duplicate comment. Did you mean to type something else?

Sorry, that was by mistake. We are considering releasing v2. There are a few changes that we would like to make before v2. We are working on it.

**Comment by jchappelow** [comment]
One more step after a v2 release is to make a `v1.0.1` that is effectively the same as `v0.2.0`.  For the most part an app can point to `v0.2.0`, unless some other dependency is already pointing to `v1.0.0` in which case you can't force that.  Also go tooling will go for `v1.0.0` when getting latest, but that's not desirable.
With both a v2 and a v1.0.1/v1.1.0 all should be well.

**Comment by mangalaman93** [comment]
We are going to retract our v1.0.0 and get a v2 out. We would like to get some changes in before we get the release out. Thank you.

**Comment by mangalaman93** [comment]
We have retracted v1.0.0 (as well as v1.0.1) and released v2.0.0-alpha https://github.com/dgraph-io/ristretto/releases/tag/v2.0.0-alpha. Let us know whether this helps. We are gonna hold for a week or so before we do the final v2.0.0. Thanks
